### PR TITLE
parse multiple relationship types

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/parser/grammar.y [hunk 1]
```diff
@@ -277,13 +278,33 @@ edge(A) ::= LEFT_BRACKET UQSTRING(B) properties(C) RIGHT_BRACKET . {
...
+}
+
+%type edgeLabels {char**}
+edgeLabels(A) ::= edgeLabel(B) . {
+	A = array_new(char*, 1);
+	A = array_append(A, B);
+}
+
+// Multiple labels
+edgeLabels(A) ::= edgeLabels(B) PIPE edgeLabel(C) . {
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: For multiple types, the Cypher standard specifies one colon for the multi-relationship format:
`(a)-[r:TYPE1|TYPE2]->(b)`
Though in practice Neo4j also accepts multiple colons like here.

File: src/graph/query_graph.c [hunk 2]
```diff
@@ -322,13 +322,10 @@ void QueryGraph_Free(QueryGraph* g) {
...
-    for(i = 0; i < nodeCount; i++) {
-        Node* n = g->nodes[i];
-        Node_Free(n);
-    }
-
-    /* TODO: Free edges. */
+    for(i = 0; i < nodeCount; i++) Node_Free(g->nodes[i]);
+    for(i = 0; i < edgeCount; i++) Edge_Free(g->edges[i]);
 
     /* Edges are freed internally by nodes. */
```
Comment (jeffreylovitz) [inline src/graph/query_graph.c]: Comment no longer accurate, I think

File: src/arithmetic/algebraic_expression.c [hunk 3]
```diff
@@ -481,16 +466,82 @@ void AlgebraicExpression_Transpose(AlgebraicExpression *ae) {
...
+        switch(exp->operation.op) {
+            case AL_EXP_MUL:
+                assert(GrB_mxm(r, NULL, NULL, Rg_structured_bool, l, r, desc) == GrB_SUCCESS);
+                break;
+            case AL_EXP_ADD:
+                assert(GrB_eWiseAdd_Matrix_Semiring(r, NULL, NULL, Rg_structured_bool, l, r, desc) == GrB_SUCCESS);
+                break;
+            case AL_EXP_TRANSPOSE:
+                // Transpose is an unary operation which gets delayed.
+                assert( !(l && r) && (l || r) ); // Verify unary.
```
Comment (jeffreylovitz) [inline src/arithmetic/algebraic_expression.c]: These conditions can be replaced with `(l ^ r)`, though I understand if you prefer what's written for readability.

File: src/arithmetic/algebraic_expression.c [hunk 4]
```diff
@@ -371,84 +380,60 @@ AlgebraicExpression **AlgebraicExpression_From_Query(const AST *ast, Vector *mat
...
+     * X = C*D
+     * Y = B*X
+     * Z = A*Y */
+    while(operand_count > 1) {
+        rightTerm = operands[operand_count-1];
+        leftTerm = operands[operand_count-2];
+
+        // Multiply and reduce.
+        if(leftTerm.transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+        if(rightTerm.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
```
Comment (jeffreylovitz) [inline src/arithmetic/algebraic_expression.c]: If we're transposing both terms, we can exploit the property:

`C = A'B'` == `C' = BA`
GraphBLAS won't apply this automatically.

File: src/arithmetic/algebraic_expression.c [hunk 5]
```diff
@@ -371,84 +380,60 @@ AlgebraicExpression **AlgebraicExpression_From_Query(const AST *ast, Vector *mat
     exp->dest_node = dest;
     AlgebraicExpression **expressions = _AlgebraicExpression_Intermidate_Expressions(exp, ast, matchPattern, q, exp_count);
     expressions = _AlgebraicExpression_IsolateVariableLenExps(expressions, exp_count);
-    AlgebraicExpression_Free(exp);
+    // AlgebraicExpression_Free(exp);
+    free(exp);
```
Comment (jeffreylovitz) [inline src/arithmetic/algebraic_expression.c]: Running the query:
`GRAPH.QUERY social "MATCH (a)-[e:friend|:visited|:new]->(b) RETURN a, b, e"`
`expressions` and `exp->operands` do not get freed.

File: src/parser/grammar.y (hunk 1 above)
Comment (swilly22) [reply src/parser/grammar.y]: I had no idea, it seems logical to me that the language should allow you to specify as many relationship types as you want

File: src/arithmetic/algebraic_expression.c (hunk 3 above)
Comment (swilly22) [reply src/arithmetic/algebraic_expression.c]: Yeah I think the long conditional is a bit more readable

File: src/arithmetic/algebraic_expression.c [hunk 6]
```diff
@@ -481,16 +466,82 @@ void AlgebraicExpression_Transpose(AlgebraicExpression *ae) {
...
+        if(leftHand && leftHand->type == AL_OPERATION && leftHand->operation.op == AL_EXP_TRANSPOSE) {
+            if(!desc) GrB_Descriptor_new(&desc);
+            GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+        }
+        
+        if(rightHand && rightHand->type == AL_OPERATION && rightHand->operation.op == AL_EXP_TRANSPOSE) {
+            if(!desc) GrB_Descriptor_new(&desc);
+            GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+        }
+
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/arithmetic/algebraic_expression.c]: On a second though we can't always do B*A
when B is a column vector [MX1] and A is a matrix [MXM]
we can do [MXM] * [MX1]
But we can't do [MX1] * [MXM]
if we were to use GrB_vxm then I think we'll be ok
Although B is really a matrix [MX1] disguising to be a vector

File: src/arithmetic/algebraic_expression.c [hunk 7]
```diff
@@ -481,16 +466,82 @@ void AlgebraicExpression_Transpose(AlgebraicExpression *ae) {
...
+    assert(root && root->type == AL_OPERATION && root->operation.l == NULL);
+    root->operation.l = child;
+}
+
+void AlgebraicExpressionNode_AppendRightChild(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child) {
+    assert(root && root->type == AL_OPERATION && root->operation.r == NULL);
+    root->operation.r = child;
+}
+
+GrB_Matrix AlgebraicExpression_Eval(const AlgebraicExpressionNode *exp) {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/arithmetic/algebraic_expression.c]: soon enough

File: src/parser/grammar.y (hunk 1 above)
Comment (jeffreylovitz) [reply src/parser/grammar.y]: Ah, sorry to be unclear! You can specify as many relationship types as you want, but the colon is only used once:
`r:TYPE1|TYPE2|TYPE3]`
is fine.

File: src/arithmetic/algebraic_expression.c (hunk 5 above)
Comment (jeffreylovitz) [reply src/arithmetic/algebraic_expression.c]: :+1:

**Pull request merged**